README.md

```diff
@@ -41,6 +41,69 @@ There are four classes you need to know about:
   message is composed of a "core" payload (memory living on CPU) plus a list of
   tensors (memory living on any device, like GPUs).
 
+Sending a message from one end of the pipe to the other can be achieved using
+the `write` method, which takes a message (with the data to send) and a
+callback which will be invoked once the sending has completed. This callback
+will be invoked with an error (if one happened) and with the message.
+
+Receiving a message takes two steps. First, you must register a callback that
+will be notified for incoming messages. This is performed by calling the
+`readDescriptor` method with said callback. The callback will be invoked with
+a so-called descriptor, which can be seen as a "message skeleton", i.e., a
+message with no buffers attached (they are set to null pointers). The job of
+this callback is filling in those buffers, either by allocating the required
+memory or by obtaining it from somehere else (from a cache, as a slice of a
+batch that's being assembled, ...). Once the message's buffers are ready, you
+can tell the pipe to go ahead and write the incoming data into them by passing
+the message to the `read` method, together with a callback which will be
+called when all the data has been received and stored. This callback, again,
+will be given a (possibly empty) error and the original message. The
+`readDescriptor` callback is one-shot, which means that after it fires it
+"expires" and will not be called again. It must be re-armed for a new event to
+be received.
+
+When you pass a message to the pipe, to send it or to receive into it, you must
+not tamper with the underlying memory until the callback has completed, even if
+the `write` or `read` call already returned. This means you can not deallocate
+the memory or alter it in any way, as the pipe may still be reading or
+modifying it. In other terms, you relinquish control over the memory when you
+pass a message to the pipe, only to reacquire it once the message is given back
+to you in the callback. This contract is encoded by the requirement to move the
+messages into and out of the pipe (using rvalue references). Also, because of
+this agreement, all callbacks will always be called, even if the pipe is closed
+or if it errors, in order to give back the memory.
+
+The order in which messages are written to a pipe is preserved when these
+messages are read on the other side. Moreover, for a given pipe endpoint, the
+callbacks of the performed operations are executed in the same order that these
```

Comment:
what are write callbacks for?
maybe it would be easier to make two sections in the README: 1) on `write` 2) on `read`

Reply:
as well as some written examples to illustrate how to use the API maybe (e.g., at the beginning of the README, tl;dr style)